packages/xstate-store/src/index.ts

```diff
@@ -0,0 +1,220 @@
+    )
+  };
+}
+
+function defaultSetter<TContext extends MachineContext>(
+  context: TContext,
+  recipe: Recipe<TContext, TContext>
+): TContext {
+  return recipe(context);
+}
+
+/**
+ * Creates a `Store` that has its own internal state and can receive events.
+ * 
+ * @example
+  ```ts
+  const store = createStore({
+    // initial context
+    { count: 0 },
+    // transitions 
+    {
+      on: {
+        inc: (context, event: { by: number }) => {
+          return {
+            count: context.count + event.by
+          }
+        }
+      }
+    }
+  });
+
+  store.getSnapshot(); // { context: { count: 0 } }
+  store.send({ type: 'inc', by: 5 });
+  store.getSnapshot(); // { context: { count: 5 } }
+  ```
+ */
+export function createStore<
+  TContext extends MachineContext,
+  TEventPayloadMap extends EventPayloadMap
+>(
+  initialContext: TContext,
+  transitions: {
+    [K in keyof TEventPayloadMap & string]:
+      | StoreAssigner<NoInfer<TContext>, { type: K } & TEventPayloadMap[K]>
+      | StorePropertyAssigner<
+          NoInfer<TContext>,
+          { type: K } & TEventPayloadMap[K]
+        >;
+  },
+  updater?: (
+    context: NoInfer<TContext>,
+    recipe: (context: NoInfer<TContext>) => NoInfer<TContext>
+  ) => NoInfer<TContext>
+): Store<TContext, ExtractEventsFromPayloadMap<TEventPayloadMap>> {
+  type StoreEvent = ExtractEventsFromPayloadMap<TEventPayloadMap>;
+  const setter = updater ?? defaultSetter;
+  let observers: Set<Observer<StoreSnapshot<TContext>>> | undefined;
+  const initialSnapshot: StoreSnapshot<TContext> = {
+    context: initialContext,
+    status: 'active'
+  };
+  let currentSnapshot = initialSnapshot;
+
+  function receive(event: StoreEvent) {
+    let currentContext = currentSnapshot.context;
+    const assigner = transitions?.[event.type as StoreEvent['type']];
+
+    if (!assigner) {
+      return;
+    }
+
+    if (typeof assigner === 'function') {
+      currentContext = updater
+        ? updater(
+            currentContext,
+            (draftContext) =>
+              (assigner as StoreCompleteAssigner<TContext, StoreEvent>)?.(
+                draftContext,
+                event
+              )
+          )
+        : setter(currentContext, (draftContext) =>
+            Object.assign(
+              {},
+              currentContext,
+              assigner?.(
+                draftContext,
+                event as any // TODO: help me
+              )
+            )
+          );
+    } else {
+      const partialUpdate: Record<string, unknown> = {};
+      for (const key of Object.keys(assigner)) {
+        const propAssignment = assigner[key];
+        partialUpdate[key] =
+          typeof propAssignment === 'function'
+            ? (
+                propAssignment as StorePartialAssigner<
+                  TContext,
+                  StoreEvent,
+                  typeof key
+                >
+              )(currentContext, event)
+            : propAssignment;
```

Comment:
this branch doesn't use `updater` at all and it should

Reply:
Just like in `@xstate/immer`, it doesn't make sense for the `produce(…)` function (the main use-case for `updater`) to operate on individual properties.

Reply:
why? 🤔 it feels odd that I would be allowed to "mutate" things in the regular assigned but I wouldn't be able to do that in per-property assigner

Reply:
Ok, i guess i see how it would be quirky but the point above still stands - it's not intuitive that a "whole" assigner works differently from the per-property assigners when u configure the whole thing with `produce`.

Reply:
What do you recommend to do? I would rather have the constraint that per-property assigners are not allowed if a producer is provided.

Reply:
Also, for per-property, you can/should just use "idiomatic immer"; same # of lines:

```ts
// No producer
addTodo: {
  todos: (context, event) => context.todos.concat(event.todo)
}

// Producer
addTodos: (context, event) => {
  context.todos.push(event.todo);
}
```

Reply:
> I would rather have the constraint that per-property assigners are not allowed if a producer is provided.

this seems to be a reasonable way out of this for now

Reply:
Ah, it's already limited like this (sort of). Nice. I pushed out a small change to make it impossible to create producer-based stores with those property assigners using `createStore` API.